core/services/headreporter/head_reporter.go

```diff
@@ -0,0 +1,129 @@
+package headreporter
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/sqlutil"
+	"github.com/smartcontractkit/chainlink/v2/core/config"
+	"github.com/smartcontractkit/chainlink/v2/core/services/telemetry"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/utils/mailbox"
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/legacyevm"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+//go:generate mockery --quiet --name HeadReporter --output ../../internal/mocks/ --case=underscore
+type (
+	HeadReporter interface {
+		ReportNewHead(ctx context.Context, head *evmtypes.Head) error
+		ReportPeriodic(ctx context.Context) error
+	}
+
+	HeadReporterService struct {
+		services.StateMachine
+		ds           sqlutil.DataSource
+		chains       legacyevm.LegacyChainContainer
+		lggr         logger.Logger
+		newHeads     *mailbox.Mailbox[*evmtypes.Head]
+		chStop       services.StopChan
+		wgDone       sync.WaitGroup
+		reportPeriod time.Duration
+		reporters    []HeadReporter
+	}
+)
+
+func NewHeadReporterService(config config.HeadReport, ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, monitoringEndpointGen telemetry.MonitoringEndpointGenerator, opts ...interface{}) *HeadReporterService {
+	reporters := make([]HeadReporter, 2)
+	reporters = append(reporters, NewPrometheusReporter(ds, chainContainer, opts))
+	if config.TelemetryEnabled() {
+		reporters = append(reporters, NewTelemetryReporter(chainContainer, lggr, monitoringEndpointGen))
+	}
+	return NewHeadReporterServiceWithReporters(ds, chainContainer, lggr, reporters, opts)
+}
+
+func NewHeadReporterServiceWithReporters(ds sqlutil.DataSource, chainContainer legacyevm.LegacyChainContainer, lggr logger.Logger, reporters []HeadReporter, opts ...interface{}) *HeadReporterService {
+	reportPeriod := 30 * time.Second
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case time.Duration:
```

Comment:
Maybe add a default case with a debug log, it might come in handy in the future

Reply:
done